hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/slowlog/TestSlowLogRecorder.java

```diff
@@ -0,0 +1,522 @@
+    Configuration conf = applySlowLogRecorderConf(14);
+    slowLogRecorder = new SlowLogRecorder(conf);
+    AdminProtos.SlowLogResponseRequest request =
+      AdminProtos.SlowLogResponseRequest.newBuilder().setLimit(14 * 11).build();
+    Assert.assertEquals(slowLogRecorder.getSlowLogPayloads(request).size(), 0);
+    LOG.debug("Initially ringbuffer of Slow Log records is empty");
+
+    for (int i = 0; i < 14 * 11; i++) {
+      RpcLogDetails rpcLogDetails =
+        getRpcLogDetails("userName_" + (i + 1), "client_" + (i + 1), "class_" + (i + 1));
+      slowLogRecorder.addSlowLogPayload(rpcLogDetails);
+    }
+    LOG.debug("Added 14 * 11 records, ringbuffer should only provide latest 14 records");
+
+    Assert.assertNotEquals(-1, HBASE_TESTING_UTILITY.waitFor(3000,
+      () -> slowLogRecorder.getSlowLogPayloads(request).size() == 14));
+
+    List<SlowLogPayload> slowLogPayloads = slowLogRecorder.getSlowLogPayloads(request);
+    Assert.assertEquals(slowLogPayloads.size(), 14);
+
+    // confirm strict order of slow log payloads
+    confirmPayloadParams(0, 154, slowLogPayloads);
+    confirmPayloadParams(1, 153, slowLogPayloads);
+    confirmPayloadParams(2, 152, slowLogPayloads);
+    confirmPayloadParams(3, 151, slowLogPayloads);
+    confirmPayloadParams(4, 150, slowLogPayloads);
+    confirmPayloadParams(5, 149, slowLogPayloads);
+    confirmPayloadParams(6, 148, slowLogPayloads);
+    confirmPayloadParams(7, 147, slowLogPayloads);
+    confirmPayloadParams(8, 146, slowLogPayloads);
+    confirmPayloadParams(9, 145, slowLogPayloads);
+    confirmPayloadParams(10, 144, slowLogPayloads);
+    confirmPayloadParams(11, 143, slowLogPayloads);
+    confirmPayloadParams(12, 142, slowLogPayloads);
+    confirmPayloadParams(13, 141, slowLogPayloads);
+
+    boolean isRingBufferCleaned = slowLogRecorder.clearSlowLogPayloads();
+    Assert.assertTrue(isRingBufferCleaned);
+    LOG.debug("cleared the ringbuffer of Online Slow Log records");
+    slowLogPayloads = slowLogRecorder.getSlowLogPayloads(request);
+    // confirm ringbuffer is empty
+    Assert.assertEquals(slowLogPayloads.size(), 0);
+  }
+
+  @Test
+  public void testOnlineSlowLogWithDefaultDisableConfig() throws Exception {
+    Configuration conf = HBASE_TESTING_UTILITY.getConfiguration();
+    conf.unset(HConstants.SLOW_LOG_BUFFER_ENABLED_KEY);
+    slowLogRecorder = new SlowLogRecorder(conf);
+    AdminProtos.SlowLogResponseRequest request =
+      AdminProtos.SlowLogResponseRequest.newBuilder().build();
+    Assert.assertEquals(slowLogRecorder.getSlowLogPayloads(request).size(), 0);
+    LOG.debug("Initially ringbuffer of Slow Log records is empty");
+
+    for (int i = 0; i < 300; i++) {
+      RpcLogDetails rpcLogDetails =
+        getRpcLogDetails("userName_" + (i + 1), "client_" + (i + 1), "class_" + (i + 1));
+      slowLogRecorder.addSlowLogPayload(rpcLogDetails);
+    }
+
+    Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
```

Comment:
Can you use waitFor() instead of sleeps?

Reply:
Make sure fix these..

Reply:
We can't use waitFor() here because we are just giving some time to ring buffer consumer to done processing all events. We don;t want to compare with any perticular result as such. Same applies to all `Uninterruptibles.sleepUninterruptibly` usecases. As per Nick's early suggestions, I had already updated `waitFor()` to all places where it fit well.
Thanks